False positive: EmptyLineSeparator reports violation on method call

reported at https://github.com/checkstyle/checkstyle/pull/18311#issuecomment-3691084557

I have read check documentation: https://checkstyle.org/checks/whitespace/emptylineseparator.html#EmptyLineSeparator
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

`EmptyLineSeparator` reports violation on method call which wrong. It only occurs when there are more than one empty lines before a method call.

#### Config:
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="EmptyLineSeparator">
      <property name="tokens"
               value="PACKAGE_DEF, IMPORT, STATIC_IMPORT, CLASS_DEF, INTERFACE_DEF, ENUM_DEF,
                    STATIC_INIT, INSTANCE_INIT, METHOD_DEF, CTOR_DEF, VARIABLE_DEF, RECORD_DEF,
                    COMPACT_CTOR_DEF"/>
      <property name="allowNoEmptyLineBetweenFields" value="true"/>
      <property name="allowMultipleEmptyLines" value="false"/>
      <property name="allowMultipleEmptyLinesInsideClassMembers" value="false"/>
    </module>
  </module>
</module>
```

#### Test.java:
```
import org.junit.Test;
import org.junit.Assert;

public class Test {
  
  public void foo(int a) {}

  @Test
  public void testFoo() {

    
    int a = 10;


    Test t = new Test();

    
    t.foo(10);
    int b = 20;


    Assert.assertFalse(false);
  }
}
```

#### Actual Cli:
```
$ java -jar checkstyle-12.3.0-all.jar -c config_check.xml Test.java
Starting audit...
[ERROR] F:\GitHub\headhtmltagname\Test.java:9:25: There is more than 1 empty line after this line. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:12:15: There is more than 1 empty line after this line. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:18:6: There is more than 1 empty line after this line. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:22:11: There is more than 1 empty line after this line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 4 errors.
```
According to the reported violation, lines 18 and 22 have more than one empty line after them, but there are no empty lines after these lines.

#### Expected Cli:
```
$ java -jar checkstyle-12.3.0-all.jar -c config_check.xml Test.java
Starting audit...
[ERROR] F:\GitHub\headhtmltagname\Test.java:9:25: There is more than 1 empty line after this line. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:12:15: There is more than 1 empty line after this line. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:15:6: There is more than 1 empty line after this line. [EmptyLineSeparator]
[ERROR] F:\GitHub\headhtmltagname\Test.java:19:11: There is more than 1 empty line after this line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 4 errors.
```
In my opinion, it might be incorrect line referencing.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java
@@ -267,18 +267,34 @@ public class EmptyLineSeparatorCheck extends AbstractCheck {
             }
         }
 
+        return positionToPotentialPostFixNode(result, line);
+    }
+
+    /**
+     * A post fix AST will always have a sibling METHOD CALL
+     * METHOD CALL will at least have two children
+     * The first child is DOT in case of POSTFIX which have at least 2 children
+     * First child of DOT again puts us back to normal AST tree which will
+     * recurse down below from here.
+     *
+     * @param postFixAst the ast to check.
+     * @param line the empty line which gives violation.
+     * @return The potential post fix node after which empty lines are present.
+     */
+    private static DetailAST positionToPotentialPostFixNode(DetailAST postFixAst, int line) {
+        DetailAST result = postFixAst;
         if (result.getNextSibling() != null) {
             final Optional<DetailAST> postFixNode = getPostFixNode(result.getNextSibling());
             if (postFixNode.isPresent()) {
-                // A post fix AST will always have a sibling METHOD CALL
-                // METHOD CALL will at least have two children
-                // The first child is DOT in case of POSTFIX which have at least 2 children
-                // First child of DOT again puts us back to normal AST tree which will
-                // recurse down below from here
                 final DetailAST firstChildAfterPostFix = postFixNode.orElseThrow();
                 result = getLastElementBeforeEmptyLines(firstChildAfterPostFix, line);
             }
         }
+
+        if (result.getLineNo() > line) {
+            result = postFixAst;
+        }
+
         return result;
     }
 

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheckTest.java
@@ -457,6 +457,19 @@ public class EmptyLineSeparatorCheckTest
     }
 
     @Test
+    public void testDisAllowMultipleEmptyLinesInsideClassMembers1() throws Exception {
+        final String[] expected = {
+            "23:27: " + getCheckMessage(MSG_MULTIPLE_LINES_INSIDE),
+            "26:19: " + getCheckMessage(MSG_MULTIPLE_LINES_INSIDE),
+            "31:64: " + getCheckMessage(MSG_MULTIPLE_LINES_INSIDE),
+            "36:19: " + getCheckMessage(MSG_MULTIPLE_LINES_INSIDE),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputEmptyLineSeparatorInsideClassMembers.java"),
+                expected);
+    }
+
+    @Test
     public void testAllowMultipleEmptyLinesInsideClassMembers() throws Exception {
         final String[] expected = {
             "53:1: " + getCheckMessage(MSG_SHOULD_BE_SEPARATED, "CLASS_DEF"),

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/emptylineseparator/InputEmptyLineSeparatorInsideClassMembers.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/emptylineseparator/InputEmptyLineSeparatorInsideClassMembers.java
@@ -1,0 +1,41 @@
+/*
+EmptyLineSeparator
+allowNoEmptyLineBetweenFields = true
+allowMultipleEmptyLines = false
+allowMultipleEmptyLinesInsideClassMembers = false
+tokens = (default)PACKAGE_DEF, IMPORT, STATIC_IMPORT, CLASS_DEF, INTERFACE_DEF, ENUM_DEF, \
+         STATIC_INIT, INSTANCE_INIT, METHOD_DEF, CTOR_DEF, VARIABLE_DEF, RECORD_DEF, \
+         COMPACT_CTOR_DEF
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.whitespace.emptylineseparator;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class InputEmptyLineSeparatorInsideClassMembers {
+
+    public void foo(int a) {}
+
+    @Test
+    public void testFoo() { // violation 'There is more than 1 empty line after this line.'
+
+
+        int a = 10; // violation 'There is more than 1 empty line after this line.'
+
+
+        // violation 2 lines below 'There is more than 1 empty line after this line.'
+        InputEmptyLineSeparatorInsideClassMembers t =
+                new InputEmptyLineSeparatorInsideClassMembers();
+
+
+        // violation 2 lines below 'There is more than 1 empty line after this line.'
+        t.foo(10);
+        int b = 20;
+
+
+        Assert.assertFalse(false);
+    }
+}

